### PR TITLE
Adding IRIG timing correction process in HWP angle calculator

### PR DESCRIPTION
**Pull request by yukisakurai**
We had IRIG timing readout issue in P10R1. This caused problems with the automatic calculations not finishing.
You can find detail info. in #319.
This PR adds this exception/correction processes to deal with this issue.
Also an option to use irig information from 2nd encoder is added.
This may be useful when the first encoder has a problem but the second encoder is working properly.
This is just optional but field_instance_sub: [instance] should be added in hwp config file to use it.

### CODE REVIEW

**Review by kmharrington — APPROVED** [review]
@yukisakurai  I made one change / commit to the code to make the second irig encoder optional. Please make sure that's what you'd like. Otherwise I was running it and getting a "referenced before assignment" error. 
  
  Otherwise this looks good to me, I'm not able to run the update script on those files from Saturday and they're producing solutions / not crashing. Although I have not checked how _good_ the solutions are, it at least looks like the automation is working.

**Review by mhasself — APPROVED** [review]
Thanks.  Is there some way that these corrections can be recorded in the output flags?  Even if you think you fixed a problem, it can be useful for someone debugging HWP angle data to see that there was (or wasn't) a correction done in some area.

File: sotodlib/hwp/g3thwp.py [hunk 1]
```diff
@@ -765,6 +806,21 @@ def _quad_form(self, quad, interp=False):
         quad[(quad > 0) & (quad < 0.5)] = 0
         return quad
 
+    def _irig_quality_check(self, irig_time, rising_edge):
+        idx = np.where(np.diff(irig_time)==1)[0]
+        if len(irig_time) == len(idx):
+            return irig_time, rising_edge
```
[the remark this answers is not in the document]
Comment (yukisakurai) [reply sotodlib/hwp/g3thwp.py:800]: Thanks, the condition was wrong here. fixed.

File: sotodlib/hwp/g3thwp.py [hunk 2]
```diff
@@ -51,10 +51,16 @@ def __init__(self, config_file=None):
...
+        
+        # 1st encoder readout 
         self._field_instance = 'observatory.HBA.feeds.HWPEncoder'
         if 'field_instance' in self.configs.keys():
             self._field_instance = self.configs['field_instance']
+        
+        # 2nd encoder readout 
+        self._field_instance_sub = None
+        if 'field_instance_sub' in self.configs.keys():
+            self._field_instance_sub = self.configs['field_instance_sub']
```
[the remark this answers is not in the document]
Comment (yukisakurai) [reply sotodlib/hwp/g3thwp.py]: All changed to `config_dict.get`.

**Comment by yukisakurai** [comment]
> Thanks. Is there some way that these corrections can be recorded in the output flags? Even if you think you fixed a problem, it can be useful for someone debugging HWP angle data to see that there was (or wasn't) a correction done in some area.

The correction process means removing wrong elements in irig_time used to extrapolate the time synchronization between the DAQ clock and the IRIG.
I added logging.debug to output removed indices.  I think it is enough info. to debugging.

**Comment by yukisakurai** [comment]
I also added some exception handlings.
Then I have confirmed that the automation is working and passed rough data quality check for at least the past two weeks. 
I will merge it to master in a few days.

**Comment by yukisakurai** [comment]
I added some exception handling and packet drop filling, and fixed the continuity of angles between data. 
Now data follow is ok and output angles are as expected except where there was an IRIG issue.
There may still be exceptions, but I recommend to use this version to analyzeP10R1 HWP data so far.
But I have "Build and Test" issue after marge master branch...hmm...

**Comment by yukisakurai** [comment]
Adding time and angle corrections in G3tHWP

**Pull request merged**